interp/src/interpreter.rs

```diff
@@ -0,0 +1,619 @@
+}
+
+// XXX(karen): I think it will need another copy of environment for each iteration of assignment statements
+/// Evaluates a group's assignment statements in an environment.
+fn eval_assigns(
+    assigns: &[ir::Assignment],
+    env: &Environment,
+) -> FutilResult<Environment> {
+    // Find the done signal in the sequence of assignments
+    let done_assign = get_done_signal(assigns);
+    // e2 = Clone the current environment
+    let mut write_env = env.clone();
+    // Get the cell that done_assign.src belongs to
+    let done_cell = get_cell_from_port(&done_assign.src);
+
+    // Prevent infinite loops; should probably be deleted later (unless we want to display the clock cycle)?
+    let mut counter = 0;
+
+    // Filter out the assignment statements that are not only from cells; for now, also excludes cells not in the env map
+    let ok_assigns = assigns
+        .iter()
+        .filter(|&a| {
+            is_cell(&a.dst.borrow())
+                && is_cell(&a.dst.borrow())
+                && env.map.contains_key(&get_cell_from_port(&a.src)) //dummy way of making sure the map has the a.src cell
+                && env.map.contains_key(&get_cell_from_port(&a.dst))
+        })
+        .collect::<Vec<_>>();
+
+    // While done_assign.src is 0 (we use done_assign.src because done_assign.dst is not a cell's port; it should be a group's port)
+    while write_env.get(&done_cell, &done_assign.src.borrow().name) == 0
+        && counter < 5
+    {
+        // println!("Clock cycle {}", counter);
+        /*println!(
+            "state of done_cell {:1} : {:?} \n",
+            &done_cell,
+            write_env.map.get(&done_cell)
+        );*/
+        // "staging" updates
+        //let mut iter_updates = write_env.clone();
+
+        // for assign in assigns
+        for assign in ok_assigns.iter() {
+            // check if the assign.guard != 0
+            // should it be evaluating the guard in write_env environment?
+            if eval_guard(&assign.guard, &write_env) {
+                // check if the cells are constants?
+                // cell of assign.src
+                let src_cell = get_cell_from_port(&assign.src);
+                // cell of assign.dst
+                let dst_cell = get_cell_from_port(&assign.dst);
+
+                /*println!(
+                    "src cell {:1} port: {:2}, dest cell {:3} port: {:4}",
+                    src_cell,
+                    &assign.src.borrow().name,
+                    dst_cell,
+                    &assign.dst.borrow().name
+                );*/
+
+                // perform a read from `env` for assign.src
+                // XXX(karen): should read from the previous iteration's env?
+                let read_val = env.get(&src_cell, &assign.src.borrow().name);
+
+                // update internal state of the cell and
+                // queue any required updates.
+
+                //determine if dst_cell is a combinational cell or not
```

Comment:
leave a comment about what happens if cell is combinational

Reply:
Added